runtime/jcl/common/jdk_internal_misc_ScopedMemoryAccess.cpp

```diff
@@ -88,21 +67,29 @@ Java_jdk_internal_misc_ScopedMemoryAccess_closeScope0(JNIEnv *env, jobject insta
 	} else {
 		vmFuncs->acquireExclusiveVMAccess(currentThread);
 		J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+		j9object_t closeScopeObj = J9_JNI_UNWRAP_REFERENCE(scope);
+#if JAVA_SPEC_VERSION >= 22
+		j9object_t errorObj = J9_JNI_UNWRAP_REFERENCE(error);
+#endif /* JAVA_SPEC_VERSION >= 22 */
 		while (NULL != walkThread) {
 			if (VM_VMHelpers::threadCanRunJavaCode(walkThread)) {
-				J9StackWalkState walkState;
-				walkState.walkThread = walkThread;
-				walkState.flags = J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_ITERATE_O_SLOTS;
-				walkState.skipCount = 0;
-				walkState.userData1 = (void *)scope;
-				walkState.userData2 = (void *)&scopeNotFound;
-				walkState.frameWalkFunction = closeScope0FrameWalkFunction;
-				walkState.objectSlotWalkFunction = closeScope0OSlotWalkFunction;
+#if JAVA_SPEC_VERSION >= 22
+				/* Skip since an exception is already pending to be thrown.*/
+				if (NULL != walkThread->scopedError) {
+					continue;
+				}
+#endif /* JAVA_SPEC_VERSION >= 22 */
 
-				vm->walkStackFrames(walkThread, &walkState);
-				if (JNI_FALSE == *(jboolean *)walkState.userData2) {
-					/* scope found */
+				if (VM_VMHelpers::hasScope(walkThread, closeScopeObj)) {
+					/* Scope found. */
+#if JAVA_SPEC_VERSION >= 22
+					setHaltFlag(walkThread, J9_PUBLIC_FLAGS_CLOSE_SCOPE);
+					walkThread->scopedError = errorObj;
+					walkThread->closeScopeObj = closeScopeObj;
```

Comment:
- Do we need to create local refs for these objects per thread so that they are not cleared by the GC? The local refs will be deleted by each thread after processing the async exception.
- Do we also need this for `J9VMThread->stopThrowable`?

Reply:
Thanks, updated.